modules/javafx.graphics/src/main/java/javafx/scene/text/Text.java

```diff
@@ -1021,27 +1021,33 @@ public final BooleanProperty caretBiasProperty() {
     public final HitInfo hitTest(Point2D point) {
         if (point == null) return null;
         TextLayout layout = getTextLayout();
+
         double x = point.getX() - getX();
         double y = point.getY() - getY() + getYRendering();
-        GlyphList[] runs = getRuns();
-        int runIndex = 0;
-        if (runs.length != 0) {
-            double ptY = localToParent(x, y).getY();
-            while (runIndex < runs.length - 1) {
-                if (ptY > runs[runIndex].getLocation().y && ptY < runs[runIndex + 1].getLocation().y) {
-                    break;
-                }
-                runIndex++;
-            }
+
+        int textRunStart = findFirstRunStart();
+
+        double px = x;
+        double py = y;
+
+        if (isSpan()) {
+            Point2D pPoint = localToParent(point);
+            px = pPoint.getX();
+            py = pPoint.getY();
         }
-        int textRunStart = 0;
-        int curRunStart = 0;
-        if (runs.length != 0) {
-            textRunStart = ((TextRun) runs[0]).getStart();
-            curRunStart = ((TextRun) runs[runIndex]).getStart();
+        TextLayout.Hit h = layout.getHitInfo((float)px, (float)py);
+        return new HitInfo(h.getCharIndex() - textRunStart, h.getInsertionIndex() - textRunStart, h.isLeading());
+    }
+
+    private int findFirstRunStart() {
+        int start = Integer.MAX_VALUE;
+        for (GlyphList r: getRuns()) {
```

Comment:
the old code had a 0 check for getRuns.length, presumably to avoid the iterator creation.
the new code is probably fine, because most of the time we'll have the need for the iterator anyway.

Reply:
`getRuns` returns an array, it won't create an `Iterator`.